R/derive_date_vars.R

```diff
@@ -101,214 +117,730 @@
+      )
+  }
+
+  # Handle min_dates and max_dates argument ----
+  restrict_imputed_dtc_dtm(
+    dtc,
+    imputed_dtc = imputed_dtc,
+    min_dates = min_dates,
+    max_dates = max_dates
+  )
+}
+
+#' Create a `dtm_level` object
+#'
+#' @param level Datetime level
+#'
+#'   *Permitted Values*: `"Y"` (year, highest level), `"M"` (month), `"D"`
+#'   (day), `"h"` (hour), `"m"` (minute), `"s"` (second, lowest level), `"n"`
+#'   (none)
+#'
+#' @returns A `dtm_level` object
+#'
+#' @details A `dtm_level` object is an ordered factor, i.e., two objects can be
+#'   compared.
+#'
+#' @author Stefan Bundfuss
+dtm_level <- function(level) {
+  out <-
+    factor(
+      level,
+      levels = c("n", "s", "m", "h", "D", "M", "Y"),
+      ordered = TRUE
     )
+  class(out) <- c("dtm_level", class(out))
+  out
+}
+
+#' Parse DTC variable and Determine Components
+#'
+#' @param dtc The `'--DTC'` date to parse
+#'
+#'   A character date is expected in a format like `yyyy-mm-dd` or
+#'   `yyyy-mm-ddThh:mm:ss`. Trailing components can be omitted and `-` is a
+#'   valid value for any component.
+#'
+#' @returns A list of character vectors. The elements of the list are named
+#'   "year", "month", "day", "hour", "minute", and "second". Missing components
+#'   are set to `NA_character_`.
+#'
+#' @details The function can be replaced by the parttime parser once it is
+#'   available.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @seealso [impute_dtc_dtm()], [impute_dtc_dt()]
+get_partialdatetime <- function(dtc) {
+  two <- "(\\d{2}|-?)"
+  partialdate <- stringr::str_match(dtc, paste0(
+    "(\\d{4}|-?)-?",
+    two,
+    "-?",
+    two,
+    "T?",
+    two,
+    ":?",
+    two,
+    ":?",
+    "(\\d{2}(\\.\\d{1,5})?)?"
+  ))
+  partial <- vector("list", 6)
+  components <- c("year", "month", "day", "hour", "minute", "second")
+  names(partial) <- components
+  for (i in seq_along(components)) {
+    partial[[i]] <- partialdate[, i + 1]
+    partial[[i]] <- if_else(partial[[i]] %in% c("-", ""), NA_character_, partial[[i]])
+  }
+  partial
+}
+
+#' Get Date Imputation Targets
+#'
+#' @param date_imputation The value to impute the day/month when a datepart is
+#'   missing.
+#'
+#'   A character value is expected, either as a
+#'   - format with month and day specified as `"mm-dd"`: e.g. `"06-15"` for the 15th
+#'   of June,
+#'   - or as a keyword: `"first"`, `"mid"`, `"last"` to impute to the first/mid/last
+#'   day/month.
+
+#' @param month Month component of the partial date
+#'
+#' @returns A list of character vectors. The elements of the list are named
+#'   "year", "month", "day".
+#'
+#' @details
+#'
+#'  - For `date_imputation = "first"` `"0000"`, `"01"`, `"01"` are returned.
+#'  - For `date_imputation = "mid"` `"xxxx"`, `"06"`, `"30"` if `month` is `NA`
+#'  and `"15"` otherwise are returned.
+#'  - For `date_imputation = "last"` `"9999"`, `"12"`, `"31"` are returned.
+#'  - For `date_imputation = "<mm>-<dd>"` `"xxxx"`, `"<mm>"`, `"<dd>"` are returned.
```

Comment:
what do the xxxx's symbolize? or is that a placeholder?

Reply:
oh is it year?  Could it be yyyy?

Reply:
"xxxx" means that it is undefined. E.g., there is no mid year. I have added a sentence.

Reply:
**ah** tyty